rust/foxglove/src/subscription/maps.rs

```diff
@@ -0,0 +1,156 @@
+//! Subscriber cache.
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use crate::channel::ChannelId;
+
+use super::SubscriberVec;
+
+/// A map of global and per-channel subscriptions.
+pub(crate) struct SubscriptionMap<K, V> {
+    /// Global subscriptions (all channels).
+    global: HashMap<K, V>,
+    /// Per-channel subscriptions.
+    channel: HashMap<ChannelId, HashMap<K, V>>,
+}
+
+impl<K, V> Default for SubscriptionMap<K, V> {
+    fn default() -> Self {
+        Self {
+            global: HashMap::default(),
+            channel: HashMap::default(),
+        }
+    }
+}
+impl<K, V> SubscriptionMap<K, V>
+where
+    K: Eq + Hash + Clone,
+    V: Clone,
+{
+    /// Removes all subscriptions.
+    pub fn clear(&mut self) {
+        self.global.clear();
+        self.channel.clear();
+    }
+
+    /// Adds a global subscription to all channels.
+    pub fn subscribe_global(&mut self, key: K, value: V) -> bool {
+        self.global.insert(key, value).is_none()
+    }
+
+    /// Adds subscriptions to the specified channels.
+    pub fn subscribe_channels(
+        &mut self,
+        key: K,
+        value: V,
+        channel_ids: impl IntoIterator<Item = ChannelId>,
+    ) -> bool {
+        let mut inserted = false;
+        for channel_id in channel_ids {
+            inserted |= self
+                .channel
+                .entry(channel_id)
+                .or_default()
```

Comment:
thank goodness this is explicit, the implicit "or default" behavior in C++ is a huge footgun 😄